core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -77,25 +80,57 @@
             if (exception[0] == null) {
                 exception[0] = e;
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
+
+            try {
+                doRollback(context, transactionManager, transaction);
+            }
+            finally {
+                rolledBack = true;
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
+                if (readOnly && !rolledBack) {
+                    try {
+                        doRollback(context, transactionManager, transaction);
+                    }
+                    finally {
+                        rolledBack = true;
+                    }
                 } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
                     doCommit(context, transactionManager, transaction);
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!rolledBack) {
+                        try {
+                            doRollback(context, transactionManager, transaction);
+                        }
+                        finally {
+                            rolledBack = true;
+                        }
+                    }
                 } else {
                     Log.debug(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
+                    if (!rolledBack) {
+                        try {
+                            doRollback(context, transactionManager, transaction);
+                        }
+                        finally {
+                            rolledBack = true;
+                        }
+                    }
                 }
             } catch (Throwable t) {
                 Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", t);
-                doRollback(context, transactionManager, transaction);
+                if (!rolledBack) {
+                    try {
+                        doRollback(context, transactionManager, transaction);
+                    }
+                    finally {
+                        rolledBack = true;
```

Comment:
This seem not used later.

Reply:
You are correct, this I think is the last one. I would leave it as written:

a) To be be clear of rollBack state (setting with try/finally after each call to doRollback

b) This way you can put a breakpoint in a debugger, or add more logic after this `catch` and be clear what is going on....

A more interesting question, if you look at the previous logic, should I set rolledBack = true *before* calling the `doRollback` method?